lib/srv/db/mysql/handshake.go

```diff
@@ -0,0 +1,308 @@
+		_, err := waitForOKPacket(serviceConn)
+		return trace.Wrap(err)
+	case handshakeWhenSupported:
+		bufConn := protocol.NewBufferedConn(ctx, serviceConn)
+		serviceConn = bufConn
+		isHandshakeV10, err := protocol.IsHandshakeV10Packet(bufConn)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if isHandshakeV10 {
+			log.DebugContext(ctx, "Received a handshake v10 packet")
+		} else {
+			log.DebugContext(ctx, "Waiting for magic OK packet from the database agent")
+			pkt, err := waitForOKPacket(serviceConn)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if !isMagicOKResult(pkt) {
+				return nil
+			}
+			if err := sendMagicProxyReply(serviceConn); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	case handshakeAlways:
+	}
+
+	log.DebugContext(ctx, "Performing a MySQL handshake with the database agent")
+	return trace.Wrap(handshakeWithAgent(ctx, serviceConn, handshakeOpts...))
+}
+
+// isMagicOKResult returns true if the OK packet received from the agent has
+// non-zero affected rows.
+// This method can be deleted after agents are updated to always initiate a
+// MySQL handshake without negotiating one with the magic OK packet.
+// TODO(gavin): DELETE IN 21.0.0
+func isMagicOKResult(packet *protocol.OK) bool {
+	return packet != nil && packet.HasAffectedRows()
+}
+
+func sendMagicProxyReply(conn net.Conn) error {
+	if _, err := conn.Write([]byte{byte(len(magicProxyReply))}); err != nil {
+		return trace.Wrap(err)
+	}
+	_, err := io.Copy(conn, strings.NewReader(magicProxyReply))
+	return trace.Wrap(err)
+}
+
+func handshakeWithAgent(ctx context.Context, serviceConn net.Conn, options ...client.Option) error {
+	_, err := client.ConnectWithDialer(ctx, "tcp", "stubAddr",
+		"stubUser",
+		emptyPassword,
+		"stubDB",
+		func(_ context.Context, _, _ string) (net.Conn, error) {
+			return serviceConn, nil
+		},
+		options...,
+	)
+	return trace.Wrap(err)
+}
+
+// notifyProxy notifies the proxy that the connection is ready on the agent side.
+func notifyProxy(ctx context.Context, log *slog.Logger, mode handshakeMode, proxyConn *server.Conn) error {
+	log = slog.With(handshakeModeLogKey, mode)
+	// no matter the path we take, the client will expect the sequence number to
+	// start with zero as we enter the command phase, and a mismatch could cause
+	// sequence number errors
+	defer proxyConn.ResetSequence()
+	switch mode {
+	case handshakeDisabled:
+		// Send back OK packet to indicate auth/connect success. At this point
+		// the original client should consider the connection phase completed.
+		log.DebugContext(ctx, "Notifying the proxy of connection success with an empty OK packet")
+		return trace.Wrap(proxyConn.WriteOK(nil))
+	case handshakeWhenSupported:
+		log.DebugContext(ctx, "Sending a magic OK packet to the proxy to request a MySQL handshake")
+		if err := proxyConn.WriteOK(newMagicOKResult()); err != nil {
+			return trace.Wrap(err)
+		}
+		proxyConn.ResetSequence()
```

Comment:
nit: Are those resets required before reading the sequence, too? I see it here, but not on the `handshakeWithProxy` function. Or are they only needed when we're leaving the connection to the real client?

(Sorry, those sequence resets are very confusing to me.)